### ISSUE TITLE
sgd.FMRegression always predicts NaN values

### ISSUE THREAD

**Issue by kuo000** [ibayer/fastFM#70]
I trained a SGD FM regression model, and when I made predictions with the model, all the predicted values are NaN. Upon a quick look on the model's learned coefficients (i.e., **w0_**, **w_**, and **V_**), most of them are NaNs.

```
from __future__ import division
from fastFM import sgd

model = sgd.FMRegression()
model.fit(Xtrain, ytrain)
model.predict(Xvalid)  # all predicted values are NaNs.

# Both Xtrain and Xvalid are CSC sparse matrices.
```

MCMC and ALS work fine and actually produce reasonably good predictions for my particular task. Any thought on possible problems with SGD?

OS: Ubuntu 16.04
Python version: 2.7.12
fastFM version: 0.2.5

**Comment by ibayer** [comment]
Try different values for the `step_size` parameter. Unfortunately SGD suffers from vanishing / exploding gradients if the `step_size` is not chosen properly. I would recommend to use the ALS solver which is more stable and usually faster.

**Comment by kuo000** [comment]
Thanks @ibayer.
A small `step_size` works. I had to shrink it to around `1e-7` for SGD to work properly. BTW, why do you think ALS is usually faster? It seems to me SGD is always faster.

**Comment by ibayer** [comment]
> BTW, why do you think ALS is usually faster? It seems to me SGD is always faster.

In my experience ALS needs less clock time to converge to the same quality as SGD even through one SGD iterations is much faster then one ALS iteration.

**Comment by kuo000** [comment]
> In my experience ALS needs less clock time to converge to the same quality as SGD even through one SGD iterations is much faster then one ALS iteration.

What counts as one ALS iteration? Is it when every model parameter (i.e., each coordinate as in general coordinate descent) gets updated once?

I guess the actual time may depend on the data size and the particular problem (e.g., is convergence intrinsically difficult). On my fairly small, sparse data (~90000 instances, ~100 effective non-zero dimension), SGD takes about 1/2~2/3 of the clock time needed by ALS to converge to comparable quality, as measured on some hold-out data. Of course, SGD needs a much bigger `n_iter` though.

**Comment by ibayer** [comment]
> What counts as one ALS iteration? Is it when every model parameter (i.e., each coordinate as in general coordinate descent) gets updated once?

That the definition that I'm using.

> I guess the actual time may depend on the data size and the particular problem (e.g., is convergence intrinsically difficult).

Absolutely, it's also easy to construct a data set where SGD converges faster. Just oversample the original data heavily. This shouldn't influence SGD at all but slow down ALS quite a bit.

> On my fairly small, sparse data (~90000 instances, ~100 effective non-zero dimension), SGD takes about 1/2~2/3 of the clock time needed by ALS to converge to comparable quality, as measured on some hold-out data. Of course, SGD needs a much bigger n_iter though.

I'm still surprised. The ALS solver is fairly optimized and carefully profiled, the SGD solver not so much.
Are you using a public data set? Can you share the experiments?

**Comment by ibayer** [comment]
That would be great.

**Comment by ibayer** [comment]
I'm closing this one as the original issue seem to be fixed after changing the sgd stepsize.
